Popover animations

maxokorokov opened the pull request:
Again this is cherry-picked and fixed/adapted from #2817.

The popover API is the same as before:

```html
<div ngbPopover="Look at me!" (shown)="..." (hidden)="..."></div>
```

`(shown)` and `(hidden)` emit synchronously as before in case of reduced motion or no animations

Will open the same thing for the tooltip once this is merged

benouat commented on src/util/popup.ts:
```diff
@@ -19,31 +25,47 @@ export class PopupService<T> {
+               {animation, runningTransition: 'continue'})
+        .pipe(tap(() => {
+          if (this._windowRef) {
+            this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._windowRef.hostView));
```
Can't we take the opportunity here to upgrade to `this._windowRef.hostView.destroy()` instead of `this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._windowRef.hostView))` ?

maxokorokov replied:
We could, sure. I wonder why it was not done before... it is already done for `this._contentRef.viewRef.destroy();` 5 lines below

maxokorokov commented:
> I assume that because of the change inside the popupService, this is why you also had to update tooltip and typeahead inside this PR ?

@benouat sure. As you can see, the popup service API has changed from:

```ts
open(...): ComponentRef<T>;
close(...): void;
```

to:

```ts
open(...): {windowRef: ComponentRef<T>, transition$: Observable<void>}
close(...): Observable<any>
```

fbasso commented on src/util/popup.ts:
```diff
@@ -19,31 +25,47 @@ export class PopupService<T> {
+    const transition$ = onStable$.pipe(
+        mergeMap(
+            () => ngbRunTransition(
+                nativeElement, ({classList}) => classList.add('show'), {animation, runningTransition: 'continue'})));
```
To be discussed, but I'm not sure but here we can use `stop` instead of `continue`. I don't know if there is a reason to not be able to revert it

The question is : do we want to behave like the alert (not reversible) or like the dialog, toast or collapse (reversible) ?

maxokorokov replied:
I would say reversible, as it works in two ways - opening/closing. I'll change that, thanks.

fbasso commented:
Thanks @maxokorokov, the code is fine for me ! 

Just two little points (in addition to @benouat comments):

- [ ] Rebase the branch on the latest commit of `animations` branch. It shouldn't be a problem, it's just to be sure that there is no problem with the #3793 changes,
- [ ] If it's now reversible, don't forget to add the corresponding tests.

maxokorokov commented:
Pushed again, integrated both @benouat's  and @fbasso's comments:
- rebased
- fixed comment typo
- using `this._windowRef.destroy()` now
- popup opening can now be reverted (@fbasso now works in the same way as modal)
- added test for opening revert → https://github.com/ng-bootstrap/ng-bootstrap/pull/3792/commits/3b2055d0e117e623fcc55f7c504b58c71c0cf5d9#diff-df50abf4356a61091d4c11859f693908R851-R880
